torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -0,0 +1,442 @@
+    } else {
+      deviceList += "," + std::to_string(device);
+    }
+  }
+  return deviceList;
+}
+
+// Get the list of devices from list of tensors
+std::vector<int> getDevicesOfTensors(const std::vector<at::Tensor>& tensors) {
+  std::vector<int> res;
+  for (auto& tensor : tensors) {
+    res.push_back(tensor.get_device());
+  }
+  return res;
+}
+
+// Helper that lets the input ncclStreams to wait for the THC stream
+void syncStreams(
+    THCState* thcState,
+    const std::vector<int>& devices,
+    std::vector<CUDAEvent>& ncclEvents,
+    std::vector<CUDAStream>& ncclStreams) {
+  CUDADevice gpuGuard;
+  for (auto i = 0; i < devices.size(); ++i) {
+    gpuGuard.setDevice(devices[i]);
+    auto currentThcStream = THCState_getCurrentStream(thcState);
+    CUDAStream& ncclStream = ncclStreams[i];
+    CUDAEvent& ncclEvent = ncclEvents[i];
+
+    C10D_CUDA_CHECK(cudaEventRecord(ncclEvent.getEvent(), currentThcStream));
+    C10D_CUDA_CHECK(
+        cudaStreamWaitEvent(ncclStream.getStream(), ncclEvent.getEvent(), 0));
+  }
+}
+
+} // namespace
+
+ProcessGroupNCCL::WorkNCCL::WorkNCCL(const std::vector<int>& devices)
+    : devices_(devices) {
+  // GPU deviceGuard
+  CUDADevice gpuGuard;
+  cudaEvents_.resize(devices.size());
+  // Now create the CUDA events
+  for (auto i = 0; i < devices.size(); ++i) {
+    // Guard the GPU
+    gpuGuard.setDevice(devices[i]);
+    cudaEvents_[i] = CUDAEvent::create(cudaEventDisableTiming);
+  }
+}
+
+ProcessGroupNCCL::WorkNCCL::~WorkNCCL() {}
+
+// Checking the work's corresponding CUDA events' status
+bool ProcessGroupNCCL::WorkNCCL::isCompleted() const {
+  CUDADevice gpuGuard;
+  for (auto i = 0; i < devices_.size(); ++i) {
+    // Get to the device
+    gpuGuard.setDevice(devices_[i]);
+    auto& cudaEvent = cudaEvents_[i];
+    // Query the event
+    auto ret = cudaEventQuery(cudaEvent.getEvent());
+    if (ret != cudaSuccess && ret != cudaErrorNotReady) {
+      // Throw directly
+      C10D_CUDA_CHECK(ret);
+    }
+    if (ret == cudaErrorNotReady) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Same as synchronize(), and will always return true
+bool ProcessGroupNCCL::WorkNCCL::wait() {
+  synchronize();
+  return true;
+}
+
+// Waiting on the work's corresponding CUDA events
+void ProcessGroupNCCL::WorkNCCL::synchronize() {
+  auto thcState = ::at::globalContext().lazyInitCUDA();
+  CUDADevice gpuGuard;
+  for (auto i = 0; i < devices_.size(); ++i) {
+    // Get to the device
+    gpuGuard.setDevice(devices_[i]);
+    auto thcStream = THCState_getCurrentStream(thcState);
+    auto& cudaEvent = cudaEvents_[i];
+    // Let THC stream wait for the NCCL stream
+    C10D_CUDA_CHECK(cudaStreamWaitEvent(thcStream, cudaEvent.getEvent(), 0));
+  }
+}
+
+bool ProcessGroupNCCL::WorkNCCL::isSuccess() const {
+  throw std::runtime_error(
+      "isSuccess() is not supported by NCCL process "
+      "group's work, isCompleted() and wait() will "
+      "either succeed or throw");
+}
```

Comment:
In this case we should just return true.

The use case here is allowing downstream to poll for completion instead of blocking. If the behavior is different for different backends it won't be possible to swap one out without rewriting code that depends on it. The code would look something like:

```
if (work->isCompleted()) {
  if (!work->isSuccess()) {
    throw work->exception();
  }
  // Do stuff
}
```

If it is not supported we can just return true and keep the API semantics the same.